Sources/zcash-swift-payment-uri/Model.swift

```diff
@@ -0,0 +1,245 @@
+    /// A list of other arbitrary key/value pairs associated with this payment.
+    let otherParams: [RequestParams]?
+}
+
+typealias RequestParams = (String, String)
+
+/// An *non-negative* decimal ZEC amount represented as specified in ZIP-321.
+/// Amount can be from 1 zatoshi (0.00000001) to the `maxSupply` of 21M ZEC (`21_000_000`)
+struct Amount {
+    enum AmountError: Error {
+        case negativeAmount
+        case greaterThanSupply
+        case tooManyFractionalDigits
+        case invalidTextInput
+    }
+    
+    static let maxFractionalDecimalDigits: Int16 = 8
+    static let decimalHandler = NSDecimalNumberHandler(
+            roundingMode: NSDecimalNumber.RoundingMode.bankers,
+            scale: Self.maxFractionalDecimalDigits,
+            raiseOnExactness: true,
+            raiseOnOverflow: true,
+            raiseOnUnderflow: true,
+            raiseOnDivideByZero: true
+    )
+
+    static let maxSupply: Decimal = 21_000_000
+
+    let value: Decimal
+    /// Initializes an Amount from a `Decimal` number
+    /// - parameter value: decimal representation of the desired amount. **Important:** `Decimal` values with more than 8 fractional digits ** will be rounded** using bankers rounding.
+    /// - returns A valid ZEC amount
+    /// - throws `Amount.AmountError` then the provided value can't represent or can't be rounded to a non-negative non-zero ZEC decimal amount.
+    init(value: Decimal) throws {
+        guard value > 0 else { throw AmountError.negativeAmount }
+
+        guard value <= Self.maxSupply else { throw AmountError.greaterThanSupply }
+
+        self.value = value
+    }
+
+    init(string: String) throws {
+        let formatter = NumberFormatter.zcashNumberFormatter
+
+        guard let decimalAmount = formatter.number(from: string)?.decimalValue else {
+            throw AmountError.invalidTextInput
+        }
+
+        guard decimalAmount.significantFractionalDecimalDigits <= Self.maxFractionalDecimalDigits else {
+            throw AmountError.tooManyFractionalDigits
+        }
+
+        try self.init(value: decimalAmount)
+    }
+
+    func toString() -> String {
+        let formatter = NumberFormatter.zcashNumberFormatter
+
+        let decimal = NSDecimalNumber(decimal: self.value)
+
+        return formatter.string(from: decimal.rounding(accordingToBehavior: Self.decimalHandler)) ?? "" // this value is already validated.
+    }
+}
+
+struct RecipientAddress {
+    let value: String
+
+    /// Initialize an opaque Recipient address that's conversible to a String with or without a validating function.
+    /// - Parameter value: the string representing the recipient
+    /// - Parameter validating: a closure that validates the given input.
+    /// - Returns: `nil` if the validating function resolves the input as invalid, or a `RecipientAddress` if the input is valid or no validating closure is passed.
+    init?(value: String, validating: ((String) -> Bool)? = nil) {
```

Comment:
This could do Bech32 validation for Sapling and Unified addresses, and base58 validation for transparent.

Reply:
These validations are left for a further iteration by composing this functionality and callers can choose how to validate these.

Reply:
According to this ZIP https://zips.z.cash/zip-0173 I could try to find a dependency that's already available for Bech32, is that correct?

Reply:
Yeah, if you can find Swift Bech32 and Base58 libraries, that would be ideal.